In [1]:
from post_procesing_tools import *
from post_procesing_tools_double import *
from test import *

### Checking degeneracies and irreps of bands

In [2]:
path_so="/home/mjocic/data_paradox/nbnd_60_sp"

C_Rso = Gather_C_so(path_so,31,form='stack')
G_Rso = Gather_G_so(path_so,31)

In [3]:
k_index = 30
E_Rso = Gather_E(path_so,units='Electronvolt')[k_index]
k_path = Gather_K(path=path_so)
k_Rpoint=k_path[k_index]

In [4]:
deg_u,deg_e = Gather_Degeneracies_so(E_Rso,functions=True)
for i in range(len(deg_e)):
    print(deg_e[i])
print(deg_e.shape)
print(deg_u)

['-17.622939431401406' '2' 'index: 0-1']
['-14.41026449708512' '4' 'index: 2-5']
['-11.861618925445676' '4' 'index: 6-9']
['-11.817305826551786' '2' 'index: 10-11']
['-10.604788938015039' '2' 'index: 12-13']
['-10.586674154278525' '4' 'index: 14-17']
['-5.579094876517392' '2' 'index: 18-19']
['-5.273322460181199' '2' 'index: 20-21']
['-3.9672568104674015' '4' 'index: 22-25']
['0.8309511009681625' '4' 'index: 26-29']
['1.0266701051883007' '2' 'index: 30-31']
['1.5756508017703224' '4' 'index: 32-35']
['1.9526585371935812' '2' 'index: 36-37']
['2.2690885863105885' '4' 'index: 38-41']
['3.8297187936032713' '2' 'index: 42-43']
['3.964876854082623' '2' 'index: 44-45']
['5.6009853125794695' '4' 'index: 46-49']
['8.191644868545142' '2' 'index: 50-51']
['8.776282073100411' '2' 'index: 52-53']
['8.994025682169028' '4' 'index: 54-57']
['10.185679468037186' '2' 'index: 58-59']
(21, 3)
[[0, 1], [2, 3, 4, 5], [6, 7, 8, 9], [10, 11], [12, 13], [14, 15, 16, 17], [18, 19], [20, 21], [22, 23, 24, 25], [

Band gap is from 3.82 eV to 3.96 eV. Bands of interest have indices 42-43,44-45 and 46-49 for 8x8 k.p spin-orbit Hamiltonian. 

In [6]:
sym_path="/home/mjocic/Post_processing/Symmetry_groups/Bilbao_server_pyreadable"
file_name="Oh_double.xml"
symmetries_double,sig_double=Gather_Symmetries_double(sym_path,file_name)

In [ ]:
for i in range(len(symmetries_double)):
    print("[{}] {}".format(i,symmetries_double[i].operation))

In [7]:
Deg_char=[]
for uk in deg_u:
    print(uk)
    k_start=uk[0]
    k_stop=uk[-1]
    char=Gather_Charachters_double(
    k_start,k_stop,symmetries_double,sig_double,
    G_Rso,C_Rso[0],C_Rso[1],k_Rpoint,details=False)
    Deg_char.append(np.array(char[:,0],dtype=float))

[0, 1]
[2, 3, 4, 5]
[6, 7, 8, 9]
[10, 11]
[12, 13]
[14, 15, 16, 17]
[18, 19]
[20, 21]
[22, 23, 24, 25]
[26, 27, 28, 29]
[30, 31]
[32, 33, 34, 35]
[36, 37]
[38, 39, 40, 41]
[42, 43]
[44, 45]
[46, 47, 48, 49]
[50, 51]
[52, 53]
[54, 55, 56, 57]
[58, 59]


In [8]:
char_g1p = np.array([1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1])
char_g1m = np.array([1,1,1,1,1,1,1,1,-1,-1,-1,-1,-1,-1,-1,-1])
char_g2p = np.array([1,1,1,-1,-1,1,1,-1,1,1,1,-1,-1,1,1,-1])
char_g2m = np.array([1,1,1,-1,-1,1,1,-1,-1,-1,-1,1,1,-1,-1,1])
char_g3p = np.array([2,2,-1,0,0,2,-1,0,2,2,-1,0,0,2,-1,0])
char_g3m = np.array([2,2,-1,0,0,2,-1,0,-2,-2,1,0,0,-2,1,0])
char_g4p = np.array([3,-1,0,-1,1,3,0,1,3,-1,0,-1,1,3,0,1])
char_g4m = np.array([3,-1,0,-1,1,3,0,1,-3,1,0,1,-1,-3,0,-1])
char_g5p = np.array([3,-1,0,1,-1,3,0,-1,3,-1,0,1,-1,3,0,-1])
char_g5m = np.array([3,-1,0,1,-1,3,0,-1,-3,1,0,-1,1,-3,0,1])

char_g6p = np.array([2,0,1,0,1.414,-2,-1,-1.414,2,0,1,0,1.414,-2,-1,-1.414])
char_g7p = np.array([2,0,1,0,-1.414,-2,-1,1.414,2,0,1,0,-1.414,-2,-1,1.414])
char_g6m= np.array([2,0,1,0,1.414,-2,-1,-1.414,-2,0,-1,0,-1.414,2,1,1.414])
char_g7m= np.array([2,0,1,0,-1.414,-2,-1,1.414,-2,0,-1,0,1.414,2,1,-1.414])

char_g8p= np.array([4,0,-1,0,0,-4,1,0,4,0,-1,0,0,-4,1,0])
char_g8m= np.array([4,0,-1,0,0,-4,1,0,-4,0,1,0,0,4,-1,0])

Nk_double = [1,6,8,12,6,1,8,6,1,6,8,12,6,1,8,6]

Oh_sinm = [[char_g1m,"$\Gamma_1^-$"],[char_g2m,"$\Gamma_2^-$"],
          [char_g3m,"$\Gamma_4^-$"],[char_g4m,"$\Gamma_4^-$"],
           [char_g5m,"$\Gamma_5^-$"]]
Oh_sinp = [[char_g1p,"$\Gamma_1^+$"],[char_g2p,"$\Gamma_2^+$"],
          [char_g3p,"$\Gamma_4^+$"],[char_g4p,"$\Gamma_4^+$"],
           [char_g5p,"$\Gamma_5^+$"]]

Oh_dubm = [[char_g6m,"$\Gamma_6^-$"],[char_g7m,"$\Gamma_7^-$"],
          [char_g8m,"$\Gamma_8^-$"]]
Oh_dubp = [[char_g6p,"$\Gamma_6^+$"],[char_g7p,"$\Gamma_7^+$"],
          [char_g8p,"$\Gamma_8^+$"]]

Oh_full = Oh_sinm + Oh_sinp + Oh_dubm + Oh_dubp

In [9]:
ir_rep=[]

for i in range(len(Deg_char)):
    for char_g in Oh_full:
        c=char_coef(char_g[0],Deg_char[i],Nk_double)
        #print(c)
        if abs(c-1.0)<1e-3:
            #print("{}\n{}".format(c,char_g))
            ir_rep.append(char_g[1])

for i in range(len(ir_rep)):
    print("{} irrep: {}".format(deg_e[i][2],ir_rep[i],Deg_char[i]))

index: 0-1 irrep: $\Gamma_6^+$
index: 2-5 irrep: $\Gamma_8^-$
index: 6-9 irrep: $\Gamma_8^-$
index: 10-11 irrep: $\Gamma_6^-$
index: 12-13 irrep: $\Gamma_7^+$
index: 14-17 irrep: $\Gamma_8^+$
index: 18-19 irrep: $\Gamma_6^-$
index: 20-21 irrep: $\Gamma_7^-$
index: 22-25 irrep: $\Gamma_8^-$
index: 26-29 irrep: $\Gamma_8^-$
index: 30-31 irrep: $\Gamma_6^-$
index: 32-35 irrep: $\Gamma_8^-$
index: 36-37 irrep: $\Gamma_7^-$
index: 38-41 irrep: $\Gamma_8^-$
index: 42-43 irrep: $\Gamma_7^-$
index: 44-45 irrep: $\Gamma_7^+$
index: 46-49 irrep: $\Gamma_8^+$
index: 50-51 irrep: $\Gamma_6^+$
index: 52-53 irrep: $\Gamma_7^+$
index: 54-57 irrep: $\Gamma_8^+$


In [10]:
char_red = char_g7p*char_g4m*char_g7m
for char_g in Oh_full:
    c=char_coef(char_g[0],char_red,Nk_double)
    if (c)!=0 and abs(c)>1e-3:
          print(char_g[1],c)

$\Gamma_1^+$ 0.9998489999999999
$\Gamma_4^+$ 1.0
$\Gamma_4^+$ 1.999849
$\Gamma_5^+$ 1.000151


In [11]:
def red_to_irreds(char_red,irred_list,Nk,present='list'):
    irreds = []
    for char_g in irred_list:
        c=char_coef(char_g[0],char_red,Nk)
        if abs(c)>1e-3:
            if present=='list': irreds.append([char_g[1],c])
            if present=='print':print(char_g[1],c)
    if present=='list': return irreds
product = red_to_irreds(char_red,Oh_full,Nk_double,'list')
print(product)

[['$\\Gamma_1^+$', 0.9998489999999999], ['$\\Gamma_4^+$', 1.0], ['$\\Gamma_4^+$', 1.999849], ['$\\Gamma_5^+$', 1.000151]]


Band gap is from 3.82 eV to 3.96 eV. Bands of interest have indices 42-43,44-45 and 46-49 for 8x8 k.p spin-orbit Hamiltonian.

They correspond to irreducible representations $\Gamma_7^-$, $\Gamma_7^+$ and $\Gamma_8^+$, respectively.

We first collect matrices for these irreps from bilbao server.

In [23]:
from test import *

path = "/home/mjocic/Post_processing/Symmetry_groups/Bilbao_server_pyreadable"

file = "GM7m_E2u_GM9.xml"
#file = "GM7m_new.xml"
GM7m = Gather_irrep2(path,file)
file_d = "GM7m_E2u_GM9+d.xml"
#file_d = "GM7m+d_new.xml"
GM7m_d = Gather_irrep2(path,file_d)
GM7m.extend(GM7m_d)

file = "GM7p_E2g_GM7.xml"
#file = "GM7p_new2.xml"
GM7p = Gather_irrep2(path,file)
file_d = "GM7p_E2g_GM7+d.xml"
#file_d = "GM7p+d_new2.xml"
GM7p_d = Gather_irrep2(path,file_d)
GM7p.extend(GM7p_d)

file = "GM8p_Fg_GM10.xml"
GM8p = Gather_irrep4(path,file)
file_d = "GM8p_Fg_GM10+d.xml"
GM8p_d = Gather_irrep4(path,file_d)
GM8p.extend(GM8p_d)

Now we make equivalent but not equal matrices from our wave-functions:
$$    
\left\langle \Psi_i | {\bf R} | \Psi_j \right\rangle = D_{ij}
$$
Where $\Psi_i$ and $\Psi_j$ are band wave-functions the partners of their corresponding irreps, ${\bf R}$ is symmetry operator and $D_{ij}$ is the matrix of the irrep. Spin wave-functions are represented as:

$$
\Psi = \psi_1 | \alpha\right\rangle + \psi_1 | \beta \right\rangle
$$

where $\psi_1$ and $\psi_2$ are coordinate dependent spin-up and spin-down wavefunctions, respectively and $|\alpha\right\rangle = (1\quad 0)^\dagger$ and $\beta \right\rangle = (0\quad 1)^\dagger$ are spin-up and spin-down operators in standard spin-representation.

In [24]:
from post_procesing_tools_double import *


path_so="/home/mjocic/data_paradox/nbnd_60_sp"
C_Rso = Gather_C_so(path_so,31,form='stack')
G_Rso = Gather_G_so(path_so,31)

k_index = 30
k_path = Gather_K(path=path_so)
k_Rpoint=k_path[k_index]

sym_path = "/home/mjocic/Post_processing/Symmetry_groups/Bilbao_server_pyreadable"
file_name="Oh_double.xml"
symmetries_double,sig_double=Gather_Symmetries_double(sym_path,file_name)

k_start=42
k_stop=43
basis_g7m = [C_Rso[0,k_start:k_stop+1],C_Rso[1,k_start:k_stop+1]]
g7m_full=Gather_Matrices_double(
    k_start,k_stop,symmetries_double,sig_double,
    G_Rso,C_Rso[0],C_Rso[1],k_Rpoint,details=True,all_elements=True)

k_start=44
k_stop=45
basis_g7p = [C_Rso[0,k_start:k_stop+1],C_Rso[1,k_start:k_stop+1]]
g7p_full=Gather_Matrices_double(
    k_start,k_stop,symmetries_double,sig_double,
    G_Rso,C_Rso[0],C_Rso[1],k_Rpoint,details=True,all_elements=True)

k_start=46
k_stop=49
basis_g8p = [C_Rso[0,k_start:k_stop+1],C_Rso[1,k_start:k_stop+1]]
g8p_full=Gather_Matrices_double(
    k_start,k_stop,symmetries_double,sig_double,
    G_Rso,C_Rso[0],C_Rso[1],k_Rpoint,details=True,all_elements=True)

Dimension of 3x3x3 square matrix is: 33
Creating G-matrix
G-matrix: Done

Now performing for identity

Symm.Op. matrix:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Spin.Op. matrix:
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]
Matrix = 
 [[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]

Now performing for 180 deg rotation - cart. axis [0,0,1]

Symm.Op. matrix:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0  1]]
Spin.Op. matrix:
[[-0.-1.j  0.+0.j]
 [ 0.+0.j  0.+1.j]]
Matrix = 
 [[ 0.        +0.32986275j -0.88304647-0.3337956j ]
 [ 0.88304647-0.3337956j   0.        -0.32986275j]]

Now performing for 180 deg rotation - cart. axis [0,1,0]

Symm.Op. matrix:
[[-1  0  0]
 [ 0  1  0]
 [ 0  0 -1]]
Spin.Op. matrix:
[[ 0.+0.j -1.+0.j]
 [ 1.+0.j  0.+0.j]]
Matrix = 
 [[ 0.        -0.9377006j  -0.34735846-0.00772579j]
 [ 0.34735846-0.00772579j  0.        +0.9377006j ]]

Now performing for 180 deg rotation - cart. axis [1,0,0]

Symm.Op. matrix:
[[ 1  0  0]
 [ 0 -1  0]
 [ 0  0 -1]]
Spin.Op. matrix:
[[ 0.+0.j -0.-1.j]
 [-0.-1.j  0.+0.j]]
Matrix = 
 [[ 0.   

Matrix = 
 [[ 0.        -0.1091245j   0.31554878-0.94261382j]
 [-0.31554878-0.94261382j  0.        +0.1091245j ]]

Now performing for inv. 120 deg rotation - cart. axis [1,1,1]

Symm.Op. matrix:
[[ 0  0 -1]
 [-1  0  0]
 [ 0 -1  0]]
Spin.Op. matrix:
[[ 0.5-0.5j -0.5-0.5j]
 [ 0.5-0.5j  0.5+0.5j]]
Matrix = 
 [[-0.5       +0.24935667j  0.77297686-0.30054621j]
 [-0.77297686-0.30054621j -0.5       -0.24935667j]]

Now performing for inv. 120 deg rotation - cart. axis [-1,1,-1]

Symm.Op. matrix:
[[ 0  0 -1]
 [ 1  0  0]
 [ 0  1  0]]
Spin.Op. matrix:
[[ 0.5+0.5j -0.5+0.5j]
 [ 0.5+0.5j  0.5-0.5j]]
Matrix = 
 [[-0.5       +0.68834392j -0.42561839+0.308272j  ]
 [ 0.42561839+0.308272j   -0.5       -0.68834392j]]

Now performing for inv. 120 deg rotation - cart. axis [1,-1,-1]

Symm.Op. matrix:
[[ 0  0  1]
 [ 1  0  0]
 [ 0 -1  0]]
Spin.Op. matrix:
[[ 0.5+0.5j  0.5-0.5j]
 [-0.5-0.5j  0.5-0.5j]]
Matrix = 
 [[-0.5       -0.35848117j -0.45742807-0.6420676j ]
 [ 0.45742807-0.6420676j  -0.5       +0.358481

Matrix = 
 [[-0.5       -0.35848117j -0.45742807-0.6420676j ]
 [ 0.45742807-0.6420676j  -0.5       +0.35848117j]]

Now performing for 120 deg rotation - cart. axis [-1,-1,1]+d

Symm.Op. matrix:
[[ 0  0 -1]
 [ 1  0  0]
 [ 0 -1  0]]
Spin.Op. matrix:
[[-0.5+0.5j -0.5-0.5j]
 [ 0.5-0.5j -0.5-0.5j]]
Matrix = 
 [[-0.5       -0.57921942j  0.11006961+0.63434182j]
 [-0.11006961+0.63434182j -0.5       +0.57921942j]]

Now performing for 120 deg rotation - cart. axis [-1,-1,-1]+d

Symm.Op. matrix:
[[0 1 0]
 [0 0 1]
 [1 0 0]]
Spin.Op. matrix:
[[-0.5-0.5j -0.5-0.5j]
 [ 0.5-0.5j -0.5+0.5j]]
Matrix = 
 [[-0.5       -0.24935667j -0.77297686+0.30054621j]
 [ 0.77297686+0.30054621j -0.5       +0.24935667j]]

Now performing for 120 deg rotation - cart. axis [-1,1,1]+d

Symm.Op. matrix:
[[ 0 -1  0]
 [ 0  0  1]
 [-1  0  0]]
Spin.Op. matrix:
[[-0.5+0.5j  0.5-0.5j]
 [-0.5-0.5j -0.5-0.5j]]
Matrix = 
 [[-0.5       +0.35848117j  0.45742807+0.6420676j ]
 [-0.45742807+0.6420676j  -0.5       -0.35848117j]]

Now perfo

Matrix = 
 [[ 0.5       -0.35848117j -0.45742807-0.6420676j ]
 [ 0.45742807-0.6420676j   0.5       +0.35848117j]]

Now performing for inv. 120 deg rotation - cart. axis [1,1,-1]+d

Symm.Op. matrix:
[[ 0 -1  0]
 [ 0  0  1]
 [ 1  0  0]]
Spin.Op. matrix:
[[-0.5-0.5j  0.5+0.5j]
 [-0.5+0.5j -0.5+0.5j]]
Matrix = 
 [[ 0.5       -0.57921942j  0.11006961+0.63434182j]
 [-0.11006961+0.63434182j  0.5       +0.57921942j]]

Now performing for inv. 120 deg rotation - cart. axis [1,-1,1]+d

Symm.Op. matrix:
[[ 0  1  0]
 [ 0  0  1]
 [-1  0  0]]
Spin.Op. matrix:
[[-0.5+0.5j -0.5+0.5j]
 [ 0.5+0.5j -0.5-0.5j]]
Matrix = 
 [[ 0.5       +0.68834392j -0.42561839+0.308272j  ]
 [ 0.42561839+0.308272j    0.5       -0.68834392j]]

Now performing for inv. 180 deg rotation - cart. axis [1,1,0]+d

Symm.Op. matrix:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0  1]]
Spin.Op. matrix:
[[-0.        -0.j          0.70710678+0.70710678j]
 [-0.70710678+0.70710678j -0.        -0.j        ]]
Matrix = 
 [[ 0.        +0.58589178j  0.46874621

Matrix = 
 [[ 0.        +0.79644965j -0.35509772-0.48946252j]
 [ 0.35509772-0.48946252j  0.        -0.79644965j]]

Now performing for 180 deg rotation - cart. axis [1,-1,0]

Symm.Op. matrix:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0 -1]]
Spin.Op. matrix:
[[ 0.        +0.j         -0.70710678+0.70710678j]
 [ 0.70710678+0.70710678j  0.        +0.j        ]]
Matrix = 
 [[ 0.        -0.52538113j -0.80710896-0.26935069j]
 [ 0.80710896-0.26935069j  0.        +0.52538113j]]

Now performing for  90 deg rotation - cart. axis [0,0,-1]

Symm.Op. matrix:
[[ 0  1  0]
 [-1  0  0]
 [ 0  0  1]]
Spin.Op. matrix:
[[0.70710678+0.70710678j 0.        +0.j        ]
 [0.        +0.j         0.70710678-0.70710678j]]
Matrix = 
 [[-0.70710678-0.21171043j  0.33352716-0.58646255j]
 [-0.33352716-0.58646255j -0.70710678+0.21171043j]]

Now performing for  90 deg rotation - cart. axis [0,0,1]

Symm.Op. matrix:
[[ 0 -1  0]
 [ 1  0  0]
 [ 0  0  1]]
Spin.Op. matrix:
[[0.70710678-0.70710678j 0.        +0.j        ]
 [0.        +0.

Matrix = 
 [[-0.70710678+0.21171043j -0.33352716+0.58646255j]
 [ 0.33352716+0.58646255j -0.70710678-0.21171043j]]

Now performing for inv.  90 deg rotation - cart. axis [-1,0,0]

Symm.Op. matrix:
[[-1  0  0]
 [ 0  0 -1]
 [ 0  1  0]]
Spin.Op. matrix:
[[0.70710678+0.j         0.        +0.70710678j]
 [0.        +0.70710678j 0.70710678+0.j        ]]
Matrix = 
 [[-0.70710678-0.66091539j -0.22600562+0.11005591j]
 [ 0.22600562+0.11005591j -0.70710678+0.66091539j]]

Now performing for inv. 180 deg rotation - cart. axis [0,1,1]

Symm.Op. matrix:
[[ 1  0  0]
 [ 0  0 -1]
 [ 0 -1  0]]
Spin.Op. matrix:
[[ 0.        +0.70710678j  0.70710678+0.j        ]
 [-0.70710678+0.j          0.        -0.70710678j]]
Matrix = 
 [[ 0.       -0.34724469j  0.9146305-0.20705595j]
 [-0.9146305-0.20705595j  0.       +0.34724469j]]

Now performing for inv. 180 deg rotation - cart. axis [0,1,-1]

Symm.Op. matrix:
[[1 0 0]
 [0 0 1]
 [0 1 0]]
Spin.Op. matrix:
[[ 0.        -0.70710678j  0.70710678+0.j        ]
 [-0.707106

Matrix = 
 [[ 0.        -0.07617617j -0.24757618-0.96586916j]
 [ 0.24757618-0.96586916j  0.        +0.07617617j]]

Now performing for  90 deg rotation - cart. axis [1,0,0]+d

Symm.Op. matrix:
[[ 1  0  0]
 [ 0  0 -1]
 [ 0  1  0]]
Spin.Op. matrix:
[[-0.70710678-0.j         -0.        +0.70710678j]
 [-0.        +0.70710678j -0.70710678-0.j        ]]
Matrix = 
 [[ 0.70710678-0.66091539j -0.22600562+0.11005591j]
 [ 0.22600562+0.11005591j  0.70710678+0.66091539j]]

Now performing for  90 deg rotation - cart. axis [0,1,0]+d

Symm.Op. matrix:
[[ 0  0  1]
 [ 0  1  0]
 [-1  0  0]]
Spin.Op. matrix:
[[-0.70710678-0.j  0.70710678-0.j]
 [-0.70710678-0.j -0.70710678-0.j]]
Matrix = 
 [[ 0.70710678-0.13553426j  0.58110334+0.3794066j ]
 [-0.58110334+0.3794066j   0.70710678+0.13553426j]]

Now performing for 180 deg rotation - cart. axis [1,0,1]+d

Symm.Op. matrix:
[[ 0  0  1]
 [ 0 -1  0]
 [ 1  0  0]]
Spin.Op. matrix:
[[-0.+0.70710678j -0.+0.70710678j]
 [-0.+0.70710678j -0.-0.70710678j]]
Matrix = 
 [[ 0. 

Matrix = 
 [[ 0.70710678-0.13553426j  0.58110334+0.3794066j ]
 [-0.58110334+0.3794066j   0.70710678+0.13553426j]]

Now performing for inv. 180 deg rotation - cart. axis [1,0,1]+d

Symm.Op. matrix:
[[ 0  0 -1]
 [ 0  1  0]
 [-1  0  0]]
Spin.Op. matrix:
[[-0.+0.70710678j -0.+0.70710678j]
 [-0.+0.70710678j -0.-0.70710678j]]
Matrix = 
 [[ 0.        -0.87262583j  0.10752154-0.47640664j]
 [-0.10752154-0.47640664j  0.        +0.87262583j]]

Now performing for inv.  90 deg rotation - cart. axis [0,-1,0]+d

Symm.Op. matrix:
[[ 0  0  1]
 [ 0 -1  0]
 [-1  0  0]]
Spin.Op. matrix:
[[-0.70710678-0.j -0.70710678-0.j]
 [ 0.70710678-0.j -0.70710678-0.j]]
Matrix = 
 [[ 0.70710678+0.13553426j -0.58110334-0.3794066j ]
 [ 0.58110334-0.3794066j   0.70710678-0.13553426j]]

Now performing for inv. 180 deg rotation - cart. axis [-1,0,1]+d

Symm.Op. matrix:
[[0 0 1]
 [0 1 0]
 [1 0 0]]
Spin.Op. matrix:
[[-0.-0.70710678j -0.+0.70710678j]
 [-0.+0.70710678j -0.+0.70710678j]]
Matrix = 
 [[ 0.        -0.44920496j -0.5

Matrix = 
 [[ 0.        +0.01672848j  0.45448834+0.23442218j -0.43092814+0.49753104j
   0.4681815 +0.29287648j]
 [-0.45448834+0.23442218j  0.        -0.70735181j -0.46306516-0.01146233j
  -0.1244813 +0.08990932j]
 [ 0.43092814+0.49753104j  0.46306516-0.01146233j  0.        +0.11418199j
  -0.58237862-0.00083248j]
 [-0.4681815 +0.29287648j  0.1244813 +0.08990932j  0.58237862-0.00083248j
   0.        +0.57644134j]]

Now performing for 180 deg rotation - cart. axis [1,-1,0]

Symm.Op. matrix:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0 -1]]
Spin.Op. matrix:
[[ 0.        +0.j         -0.70710678+0.70710678j]
 [ 0.70710678+0.70710678j  0.        +0.j        ]]
Matrix = 
 [[ 0.        -0.8992097j  -0.29076547+0.018864j   -0.28815504+0.04533771j
   0.07146704-0.12776987j]
 [ 0.29076547+0.018864j    0.        +0.09714436j -0.21399094+0.56731162j
  -0.71451095+0.16583749j]
 [ 0.28815504+0.04533771j  0.21399094+0.56731162j  0.        +0.46757165j
   0.50300145-0.27503249j]
 [-0.07146704-0.12776987j  0.7145109

Matrix = 
 [[ 0.        +0.35292419j -0.69511587-0.4669257j  -0.14780768+0.3144875j
   0.18278929+0.14169481j]
 [ 0.69511587-0.4669257j   0.        -0.29614803j -0.15396843-0.22477521j
   0.36612166+0.05306055j]
 [ 0.14780768+0.3144875j   0.15396843-0.22477521j  0.        +0.01238456j
   0.07579068-0.89393665j]
 [-0.18278929+0.14169481j -0.36612166+0.05306055j -0.07579068-0.89393665j
   0.        -0.06916071j]]

Now performing for inv. 180 deg rotation - cart. axis [1,0,0]

Symm.Op. matrix:
[[-1  0  0]
 [ 0  1  0]
 [ 0  0  1]]
Spin.Op. matrix:
[[ 0.+0.j -0.-1.j]
 [-0.-1.j  0.+0.j]]
Matrix = 
 [[ 0.        +0.20805433j -0.12910299+0.14116925j -0.45445716-0.41996844j
   0.33024257-0.65433332j]
 [ 0.12910299+0.14116925j  0.        +0.59604764j -0.46094537-0.14610043j
   0.04508529+0.61014935j]
 [ 0.45445716-0.41996844j  0.46094537-0.14610043j  0.        -0.58536504j
  -0.20072005+0.0183973j ]
 [-0.33024257-0.65433332j -0.04508529+0.61014935j  0.20072005+0.0183973j
   0.        -0.21873693

Matrix = 
 [[ 0.08107149+0.11994119j -0.4779389 +0.765415j    0.38355443+0.07263205j
   0.00519407+0.11108292j]
 [-0.36422452+0.31331142j -0.18694617-0.04999403j -0.11602129-0.43023136j
   0.72093168+0.11586943j]
 [ 0.38717437+0.57367026j  0.26670453-0.16536894j  0.20736644+0.42712805j
   0.23135967+0.37888488j]
 [-0.43733606+0.27920172j  0.16335078-0.15200112j  0.64955273+0.04114194j
  -0.10149176-0.49707521j]]

Now performing for inv. 180 deg rotation - cart. axis [0,1,1]

Symm.Op. matrix:
[[ 1  0  0]
 [ 0  0 -1]
 [ 0 -1  0]]
Spin.Op. matrix:
[[ 0.        +0.70710678j  0.70710678+0.j        ]
 [-0.70710678+0.j          0.        -0.70710678j]]
Matrix = 
 [[ 0.        -0.0069256j   0.09954569-0.05704524j  0.19794655-0.59023951j
   0.58770241+0.50381435j]
 [-0.09954569-0.05704524j  0.        -0.72910742j -0.4728286 +0.15709073j
  -0.12855105+0.43642768j]
 [-0.19794655-0.59023951j  0.4728286 +0.15709073j  0.        +0.48377736j
   0.35385937+0.07023614j]
 [-0.58770241+0.50381435j  0.128

Matrix = 
 [[ 0.0746892 -0.24279143j  0.66646739+0.46878656j  0.07972194-0.41571863j
  -0.26653074-0.14600691j]
 [-0.15124854-0.17410939j -0.20331534+0.51776783j -0.60567585+0.12164713j
  -0.12134526+0.49094114j]
 [-0.64918573+0.34040651j -0.066383  +0.08545445j  0.46480629-0.18056205j
  -0.27991893+0.35209372j]
 [-0.43303277-0.39679663j -0.12212153+0.00410849j -0.03027455-0.43545044j
   0.66381986-0.09441435j]]

Now performing for 120 deg rotation - cart. axis [-1,-1,1]+d

Symm.Op. matrix:
[[ 0  0 -1]
 [ 1  0  0]
 [ 0 -1  0]]
Spin.Op. matrix:
[[-0.5+0.5j -0.5-0.5j]
 [ 0.5-0.5j -0.5-0.5j]]
Matrix = 
 [[-0.12996333-0.28691179j  0.11596236-0.11342253j  0.4344788 +0.3217622j
  -0.40926015+0.64395855j]
 [-0.71924926+0.45511946j  0.25367841+0.07059757j -0.02016591+0.42621145j
   0.0384188 -0.15056943j]
 [-0.13030395-0.25182617j -0.55614398-0.2373022j   0.49137617+0.26307951j
   0.09205356-0.48462436j]
 [ 0.29384149-0.09797741j  0.23410027+0.69565789j  0.34389328+0.31098303j
   0.38490874-0.

Matrix = 
 [[-0.08107149+0.11994119j  0.36422452+0.31331142j -0.38717437+0.57367026j
   0.43733606+0.27920172j]
 [ 0.4779389 +0.765415j    0.18694617-0.04999403j -0.26670453-0.16536894j
  -0.16335078-0.15200112j]
 [-0.38355443+0.07263205j  0.11602129-0.43023136j -0.20736644+0.42712805j
  -0.64955273+0.04114194j]
 [-0.00519407+0.11108292j -0.72093168+0.11586943j -0.23135967+0.37888488j
   0.10149176-0.49707521j]]

Now performing for  90 deg rotation - cart. axis [0,1,0]+d

Symm.Op. matrix:
[[ 0  0  1]
 [ 0  1  0]
 [-1  0  0]]
Spin.Op. matrix:
[[-0.70710678-0.j  0.70710678-0.j]
 [-0.70710678-0.j -0.70710678-0.j]]
Matrix = 
 [[ 0.008835  +0.0100302j   0.02399907-0.05550931j -0.13079356+0.53514845j
  -0.60502174-0.57150851j]
 [-0.33408918+0.27072925j -0.36977592+0.04110457j -0.05956846-0.67959462j
  -0.40906513-0.20960617j]
 [ 0.49229743+0.58699455j  0.11754175+0.31253169j  0.45942148-0.06625538j
   0.09672282-0.27708363j]
 [-0.44087559+0.18335969j  0.86351864+0.02761119j  0.01030745-0.124

Matrix = 
 [[ 0.0746892 +0.24279143j -0.15124854+0.17410939j -0.64918573-0.34040651j
  -0.43303277+0.39679663j]
 [ 0.66646739-0.46878656j -0.20331534-0.51776783j -0.066383  -0.08545445j
  -0.12212153-0.00410849j]
 [ 0.07972194+0.41571863j -0.60567585-0.12164713j  0.46480629+0.18056205j
  -0.03027455+0.43545044j]
 [-0.26653074+0.14600691j -0.12134526-0.49094114j -0.27991893-0.35209372j
   0.66381986+0.09441435j]]

Now performing for inv. 120 deg rotation - cart. axis [1,1,-1]+d

Symm.Op. matrix:
[[ 0 -1  0]
 [ 0  0  1]
 [ 1  0  0]]
Spin.Op. matrix:
[[-0.5-0.5j  0.5+0.5j]
 [-0.5+0.5j -0.5+0.5j]]
Matrix = 
 [[-0.12996333+0.28691179j -0.71924926-0.45511946j -0.13030395+0.25182617j
   0.29384149+0.09797741j]
 [ 0.11596236+0.11342253j  0.25367841-0.07059757j -0.55614398+0.2373022j
   0.23410027-0.69565789j]
 [ 0.4344788 -0.3217622j  -0.02016591-0.42621145j  0.49137617-0.26307951j
   0.34389328-0.31098303j]
 [-0.40926015-0.64395855j  0.0384188 +0.15056943j  0.09205356+0.48462436j
   0.3849087

Matrix = 
 [[ 0.008835  -0.0100302j  -0.33408918-0.27072925j  0.49229743-0.58699455j
  -0.44087559-0.18335969j]
 [ 0.02399907+0.05550931j -0.36977592-0.04110457j  0.11754175-0.31253169j
   0.86351864-0.02761119j]
 [-0.13079356-0.53514845j -0.05956846+0.67959462j  0.45942148+0.06625538j
   0.01030745+0.1246909j ]
 [-0.60502174+0.57150851j -0.40906513+0.20960617j  0.09672282+0.27708363j
  -0.09848056-0.0151206j ]]

Now performing for inv. 180 deg rotation - cart. axis [-1,0,1]+d

Symm.Op. matrix:
[[0 0 1]
 [0 1 0]
 [1 0 0]]
Spin.Op. matrix:
[[-0.-0.70710678j -0.+0.70710678j]
 [-0.+0.70710678j -0.+0.70710678j]]
Matrix = 
 [[ 0.        +0.88081321j  0.42440946-0.06334691j  0.18335646-0.03547024j
  -0.06417065+0.03219106j]
 [-0.42440946-0.06334691j  0.        -0.84721282j -0.03408345+0.03762096j
   0.003676  +0.3090367j ]
 [-0.18335646-0.03547024j  0.03408345+0.03762096j  0.        -0.95751227j
   0.20967145-0.04187337j]
 [ 0.06417065+0.03219106j -0.003676  +0.3090367j  -0.20967145-0.041873

In [25]:
r_g7m = Compute_small_r(GM7m, g7m_full)
r_g7p = Compute_small_r(GM7p, g7p_full)
r_g8p = Compute_small_r(GM8p, g8p_full)

U_g7m = Gather_Unitary_Transforms(r_g7m, GM7m, g7m_full)[0,0]
U_g7p = Gather_Unitary_Transforms(r_g7p, GM7p, g7p_full)[0,0]
U_g8p = Gather_Unitary_Transforms(r_g8p, GM8p, g8p_full)[0,0]

In [26]:
k_start=42
k_stop=43
basis_g7m = np.array([C_Rso[0,k_start:k_stop+1],C_Rso[1,k_start:k_stop+1]])
k_start=44
k_stop=45
basis_g7p = np.array([C_Rso[0,k_start:k_stop+1],C_Rso[1,k_start:k_stop+1]])
k_start=46
k_stop=49
basis_g8p = np.array([C_Rso[0,k_start:k_stop+1],C_Rso[1,k_start:k_stop+1]])

In [27]:
psi_up,psi_dn = U_g7m.T.dot(basis_g7m[0]), U_g7m.T.dot(basis_g7m[1])
newbasis_g7m = np.array([psi_up,psi_dn])

psi_up,psi_dn = U_g7p.T.dot(basis_g7p[0]), U_g7p.T.dot(basis_g7p[1])
newbasis_g7p = np.array([psi_up,psi_dn])

psi_up,psi_dn = U_g8p.T.dot(basis_g8p[0]), U_g8p.T.dot(basis_g8p[1])
newbasis_g8p = np.array([psi_up,psi_dn])

In [28]:
newbasis = newbasis_g7p
basis = basis_g7p
m,n = 1,1
A_up,B_up=newbasis[0,m],newbasis[0,n]
A_dn,B_dn=newbasis[1,m],newbasis[1,n]
print(np.vdot(A_up,B_up)+np.vdot(A_dn,B_dn))
A_up,B_up=basis[0,m],basis[0,n]
A_dn,B_dn=basis[1,m],basis[1,n]
print(np.vdot(A_up,B_up)+np.vdot(A_dn,B_dn))

(1.0000000000022122+0j)
(0.9999999999999956+0j)


In [29]:
spin_up=np.concatenate((newbasis_g7m[0],newbasis_g7p[0],newbasis_g8p[0]),axis=0)
spin_dn=np.concatenate((newbasis_g7m[1],newbasis_g7p[1],newbasis_g8p[1]),axis=0)
basis8 = np.array([spin_up,spin_dn])
print(basis8.shape)

(2, 8, 16560)


We compute matrix elements of $\left\langle {\bf p} \right\rangle$:

In [30]:
#for m in range(len(basis8[0])):
m=0
for n in range(len(basis8[0])):
    up = Compute_Matrix_Element(G_Rso,basis8[0],m,n)
    dn = Compute_Matrix_Element(G_Rso,basis8[1],m,n)
    print("KdP_so[{},{}] =\n{}\n".format(m,n,up+dn))

KdP_so[0,0] =
[-0.49999994+3.73523745e-24j -0.49999987+1.65436123e-24j
 -0.50000002-1.08416753e-19j]

KdP_so[0,1] =
[-1.51290926e-07+3.10543946e-08j -2.43893905e-07-7.28143244e-08j
 -1.74847459e-07-1.12018011e-08j]

KdP_so[0,2] =
[-1.60148694e-07-2.26474107e-07j -1.00015943e-07+5.78244231e-08j
 -3.80866921e-01+1.26040675e-01j]

KdP_so[0,3] =
[-3.80866994e-01+1.26040778e-01j  1.26040783e-01+3.80867099e-01j
 -1.57331826e-08-1.29702016e-07j]

KdP_so[0,4] =
[ 2.57359897e-09+2.12950377e-07j -1.61615038e-07+3.10031833e-07j
 -2.55986978e-01-1.93827088e-01j]

KdP_so[0,5] =
[-3.98657116e-02+3.18604912e-01j -1.24777858e-01-2.95852639e-01j
  1.56523287e-07-6.18210393e-08j]

KdP_so[0,6] =
[-1.20968356e-01+2.97430945e-01j -2.53477188e-01+1.97098464e-01j
  4.71908022e-09+7.81844671e-08j]

KdP_so[0,7] =
[ 1.68299933e-07-2.34527936e-07j  3.03331011e-07+6.36473471e-09j
 -4.39539103e-02-3.18066747e-01j]



And correct the phases of wave-functions to reduce number of parameters:

In [31]:
A = -3.80866921e-01+1.26040675e-01j
B = -2.55986978e-01-1.93827088e-01j
G = -4.39539103e-02-3.18066747e-01j
phi_a = np.arctan(A.imag/A.real)
B = B*np.exp(-1j*phi_a)
phi_b = -np.arctan(B.imag/B.real)
G = G*np.exp(-1j*phi_a)
phi_g = -np.arctan(G.imag/G.real)

basis8[:,0:2] = np.exp(1j*phi_a)*basis8[:,0:2]
basis8[:,4:6] = np.exp(1j*phi_b)*basis8[:,4:6]
basis8[:,6:8] = np.exp(1j*phi_g)*basis8[:,6:8]
#z_a = A.real*(np.cos(phi) + np.sin(phi)**2/np.cos(phi))
#print(abs(A),abs(z_a))
#B = [1,2,3,4,5,6]
#B = np.array(B)
#print(B*np.exp(1j*phi))

Check parameters now:

In [32]:
#for m in range(len(basis8[0])):
m=0
for n in range(len(basis8[0])):
    up = Compute_Matrix_Element(G_Rso,basis8[0],m,n)
    dn = Compute_Matrix_Element(G_Rso,basis8[1],m,n)
    print("KdP_so[{},{}] =\n{}\n".format(m,n,up+dn))

KdP_so[0,0] =
[-0.49999994+3.38848205e-20j -0.49999987+6.76964613e-21j
 -0.50000002+6.13923111e-25j]

KdP_so[0,1] =
[-1.51290926e-07+3.10543946e-08j -2.43893905e-07-7.28143244e-08j
 -1.74847459e-07-1.12018010e-08j]

KdP_so[0,2] =
[-8.08872411e-08-2.65321276e-07j -1.13118618e-07+2.34740496e-08j
 -4.01180587e-01-1.07475053e-10j]

KdP_so[0,3] =
[-4.01180689e-01+7.49440627e-08j  4.70135040e-08+4.01180790e-01j
  2.58125170e-08-1.28077554e-07j]

KdP_so[0,4] =
[ 1.30600343e-07+1.68220204e-07j  5.83053632e-08+3.44731262e-07j
 -3.21089198e-01+4.66356623e-10j]

KdP_so[0,5] =
[ 1.60544670e-01+2.78071527e-01j -2.78071523e-01-1.60544467e-01j
  8.74689385e-08-1.43772616e-07j]

KdP_so[0,6] =
[-2.78071657e-01-1.60544898e-01j -1.60544552e-01-2.78071818e-01j
 -7.80944536e-08-6.02801148e-09j]

KdP_so[0,7] =
[ 2.09281583e-07+1.98820118e-07j -4.78277843e-08+2.99604264e-07j
  3.21089398e-01-3.38684742e-11j]



We can now see that we have less parameters and some lost their complex part. In total we have 3 parameters:

$P_0 =$ -4.01180793e-01,

$M =$ -3.21089198e-01 and

$N =$ 1.60544670e-01+2.78071527e-01j

In [33]:
basis8_qe = np.array([C_Rso[0,42:50],C_Rso[1,42:50]])

path_so_more="/home/mjocic/data_paradox/nbnd_240_sp"
C_Rso = Gather_C_so(path_so_more,31,form='stack')
G_Rso = Gather_G_so(path_so_more,31)

basis_ext_up = np.vstack((C_Rso[0,:42],C_Rso[0,50:]))
basis_ext_dn = np.vstack((C_Rso[1,:42],C_Rso[1,50:]))


E_Rso = Gather_E(path_so_more,units='Hartree')[k_index]
Energy8 = E_Rso[42:50]
E_ext = np.append(E_Rso[:42],E_Rso[50:])

In [34]:
k_mesh = k_path[20:40]

h0 , h0_eig = Compute_H0(k_mesh,k_Rpoint,Energy8,a=10.8)
kdp1, kdp1_eig = Compute_KdP1_spin(k_mesh,k_Rpoint,
                                   G_Rso,basis8[0],basis8[1],a=10.8)
kdp2, kdp2_eig = Compute_KdP2_spin(k_mesh,k_Rpoint,Energy8,E_ext,
                                   G_Rso,basis8[0],basis8[1],
                                   basis_ext_up,basis_ext_dn,a=10.8)
#relkin, relkin_eig = Compute_KdP_RelKin(k_mesh,k_Rpoint,
#                                   G_Rso,basis8[0],basis8[1],a=10.8)
    

#h0_qe , h0_eig_qe = Compute_H0(k_mesh,k_Rpoint,Energy8,a=10.8)
#kdp1_qe, kdp1_eig_qe = Compute_KdP1_spin(k_mesh,k_Rpoint,
#                                   G_Rso,basis8_qe[0],basis8_qe[1],a=10.8)
#kdp2_qe, kdp2_eig_qe = Compute_KdP2_spin(k_mesh,k_Rpoint,Energy8,E_ext,
#                                   G_Rso,basis8_qe[0],basis8_qe[1],
#                                   basis_ext_up,basis_ext_dn,a=10.8)
#relkin, relkin_eig = Compute_KdP_RelKin(k_mesh,k_Rpoint,
#                                   G_Rso,basis8[0],basis8[1],a=10.8)

Calculating H0: DONE
Calculating KdP1: DONE

Calculating KdP2: DONE



In [35]:
H_1 = h0 + kdp1 ;        eig_h1 = np.linalg.eigvalsh(H_1)
H_2 = H_1 + kdp2 ;       eig_h2 = np.linalg.eigvalsh(H_2)
#H_3 = H_2 + relkin;      eig_h3 = np.linalg.eigvalsh(H_3)
#H_1_qe = h0_qe + kdp1_qe ;        eig_h1_qe = np.linalg.eigvalsh(H_1_qe)
#H_2_qe = H_1_qe + kdp2_qe ;       eig_h2_qe = np.linalg.eigvalsh(H_2_qe)

In [36]:
k2=[0.0]
d=0
k_mesh4=k_mesh
for i in range(len(k_mesh4)-1):
    d1 = k_mesh4[i+1]-k_mesh4[i]
    d2 = np.sqrt(np.vdot(d1,d1))
    d += d2
    k2.append(d)
k2=np.array(k2)

#path = '/home/mjocic/Desktop/plot_tmp'
units = 1#*2*13.605698066
Edft = Gather_E(path_so,units='Hartree')[20:40,42:50]

#eig_0=eig_h3 * units
eig_1=eig_h2 *units
#eig_2=eig_h2_qe *units
x=[];y_0=[];y_dft=[];y_1=[];y_2=[]
#FILE_kp = open( path+'/plotband_h012.dat','w')

for n in range(eig_1.shape[1]):
    for k_i in range(len(k_mesh4)):
        x.append(k2[k_i])
        #y_0.append(eig_0[k_i][n].real);
        y_1.append(eig_1[k_i][n].real)
        #y_2.append(eig_2[k_i][n].real)
        y_dft.append(Edft[k_i][n])
        #FILE_kp.write("{} {} {} {} {}\n".format (k2[k_i],eig_0[k_i][n].real,eig_1[k_i][n].real,eig_2[k_i][n].real,Edft[k_i][n]))
    #FILE_kp.write("\n")
    #FILE_dft.write("\n")
#FILE_kp.close()
#FILE_dft.close()

In [41]:
import matplotlib.pyplot as plt
%matplotlib
#plt.plot(x,y_0,label='$H_2+relkin$')
plt.plot(x,y_1,'x',label='$H_2$')
#plt.plot(x,y_2,'+',label='$H_2$')
plt.plot(x,y_dft,'p' ,label='dft')
#plt.xlim(k2[0],k2[-1])
#plt.ylim(y_dft[0],y_dft[1500])
#plt.ylim(-0.2,0.3)
#plt.xticks([k2[0],k2[10],k2[19]],('X','R','M'))
plt.ylabel('E(k) [Hartree]')
plt.legend()
plt.show()

Using matplotlib backend: Qt5Agg


In [39]:
def Compute_Perturbation_mat_so(G_vec,C_terms_up,C_terms_dn,
                                ext_basis_up,ext_basis_dn,
                                E_terms,E_ext,m,n):
        new_basis_up = np.vstack((ext_basis_up,C_terms_up))
        new_basis_dn = np.vstack((ext_basis_dn,C_terms_dn))
        dim = len(ext_basis_up)
        cgc = 0
        mat = np.zeros((3,3),dtype=complex)
        Eavg = (E_terms[m]+E_terms[n])/2
        for l in range(len(ext_basis_up)):
            delta_E1 =1/(Eavg - E_ext[l])
            kpi_up = Compute_Matrix_Element(G_vec,new_basis_up,dim+m,l)
            kpi_dn = Compute_Matrix_Element(G_vec,new_basis_dn,dim+m,l)
            kpj_up = Compute_Matrix_Element(G_vec,new_basis_up,l,dim+n)
            kpj_dn = Compute_Matrix_Element(G_vec,new_basis_dn,l,dim+n)
            for i in range(3):
                for j in range(3):
                    uu = kpi_up[i]*kpj_up[j]
                    dd = kpi_dn[i]*kpj_dn[j]
                    ud = kpi_up[i]*kpj_dn[j]
                    du = kpi_dn[i]*kpj_up[j]
                    mat_ij = uu+dd+ud+du
                    mat[i,j] += mat_ij*delta_E1
        return mat

In [40]:
for m in range(len(basis8[0])):
    for n in range(len(basis8[0])):
        mat_pipj = Compute_Perturbation_mat_so(G_Rso,
                                               basis8[0],basis8[1],
                                               basis_ext_up,basis_ext_dn,
                                               Energy8,E_ext,m,n)
        print("\n[{},{}]:\n".format(m,n))
        for a in range(len(mat_pipj)):
            for b in range(len(mat_pipj)):
                ab=mat_pipj[a,b]
                ba=mat_pipj[b,a]
                summ=ab+ba
                if (abs(ab)>1e-3) and (abs(summ)>1e-4):
                    print("\t[{},{}]={}".format(a,b,mat_pipj[a,b]))
        


[0,0]:

	[0,0]=(-2.456298209113401+7.762075754258971e-18j)
	[1,1]=(-2.4562986116433856-2.0814265136296656e-17j)
	[2,2]=(-2.4562983230464432-7.647775466587071e-18j)

[0,1]:


[0,2]:


[0,3]:


[0,4]:


[0,5]:


[0,6]:


[0,7]:


[1,0]:


[1,1]:

	[0,0]=(-2.4562982288527175+3.279552708147154e-17j)
	[1,1]=(-2.456298821970893-7.589390574525018e-18j)
	[2,2]=(-2.4562982366530393-5.329521952978118e-18j)

[1,2]:


[1,3]:


[1,4]:


[1,5]:


[1,6]:


[1,7]:


[2,0]:


[2,1]:


[2,2]:

	[0,0]=(3.5884680534992137+8.329571077312647e-17j)
	[1,1]=(3.5884683608698165+1.5123465604815635e-16j)
	[2,2]=(3.5884669905633118-1.413623056518666e-17j)

[2,3]:


[2,4]:

	[0,0]=(-1.1490808051591845-1.9902668634571892j)
	[0,1]=(1.3893997377124803+0.28773204611640374j)
	[1,0]=(1.3893997341949125-0.28773225225981486j)
	[1,1]=(-1.1490787903016952+1.9902670600769081j)
	[2,2]=(2.298161619966475-2.0839735982181796e-06j)

[2,5]:

	[0,2]=(1.3471230118010251+0.4455158478759037j)
	[1,2]=(-0.4455166001467584-1.347122214976

In [ ]:
pl = 1.7992206988924978e-07+0.42377111215346913j
lp = 1.7992206989488764e-07-0.423771112153469j
s = pl+lp
print(abs(s)>1e-5)

In [ ]:
1.3893997377124803+1.3893997341949125